### PR TITLE
[qtbase] add namespace support

### LINKED ISSUE #22712: [Qt] Support for `-DQT_NAMESPACE=XXX`
Hi,

Qt offers the option to put all the classes in a namespace.
For that we'd need to pass `-DQT_NAMESPACE=XXX`.

I haven't seen a way to parametrize a feature with a string. Did I miss something?

Two options:
1. Generate a random namespace based upon user+host+date
2. pick the namespace name from `QT_NAMESPACE` which would be defined in the triplet

What do you think is the right approach?

### PR DESCRIPTION
**Pull request by abique**
Fixes #22712

### CODE REVIEW

**Comment by dg0yt** [comment]
This cannot be done as a feature: It changes source compatibility. A consuming port which is not prepared for Qt with namespaces will break when another port requires Qt with namespaces.

**Comment by abique** [comment]
> This cannot be done as a feature: It changes source compatibility. A consuming port which is not prepared for Qt with namespaces will break when another port requires Qt with namespaces.

So we just check if QT_NAMESPACE is exported?
The point of the feature is to make it discoverable, but the feature is not part of the defaults and needs QT_NAMESPACE to be set in the triplet to actually take effect.

**Comment by abique** [comment]
If you build qt with a namespace, anything linking to `Qt6::Core` will inherit the compile definition `-DQT_NAMESPACE=XXX`.

Anyway if no feature and triplet only is good for everyone I'm fine with it too.

**Comment by dg0yt** [comment]
> The point of the feature is to make it discoverable.

This is what I assumed. However, the side effects are not so obvious.

> If you build qt with a namespace, anything linking to `Qt6::Core` will inherit the compile definition `-DQT_NAMESPACE=XXX`.

The problem is that forward declarations need to be wrapped by special macros.

**Comment by abique** [comment]
I'd like to add that it is really important for me to be able to set the namespace because if you implement a plugin statically linked to qt on mac, you may end up having symbol clashes with the host or other plugins using Qt due to the few objc classes which rely upon a flat namespace, and the only solution I've found is QT_NAMESPACE.

```
objc[45644]: Class QNSApplication is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fb58) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110bfffe0). One of the two will be used. Which one is undefined.
objc[45644]: Class QCocoaApplicationDelegate is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fba8) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110c00030). One of the two will be used. Which one is undefined.
objc[45644]: Class QNSColorPanelDelegate is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fbf8) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110c00080). One of the two will be used. Which one is undefined.
... (and so on...)```

**Comment by abique** [comment]
> The problem is that forward declarations need to be wrapped by special macros.

True, but if they aren't it is a programing mistake. They should be wrapped by `QT_NAMESPACE_BEGIN/END` (or something similar).

So I don't see it as a valid argument to forbid the feature.

Also you don't specify `set(QT_NAMESPACE MyQt)` in your triplet by mistake. So it is likely that you really **need** it, and know what you are doing.

**Comment by dg0yt** [comment]
I'm not arguing that supporting Qt's namespace feature is wrong, or that not using the macros is right. The point is:

- There are packages and user projects which don't use the macros. If they build without the *vcpkg feature* enabled, they shall not break with the *vcpkg feature* enabled.
- If the feature alone doesn't change anything, it is pointless. It is properly solved by a triplet variable.

It is just the combination of feature and triplet variable which has the potential to create bad user experience, and it can be easily avoided.

Why add not the triplet variable hint to the port's main description?

**Comment by abique** [comment]
> Why not add the triplet variable hint to the port's main description?

I'm happy with that solution.

**Comment by Neumann-A** [comment]
```
objc[45644]: Class QNSApplication is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fb58) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110bfffe0). One of the two will be used. Which one is undefined.
objc[45644]: Class QCocoaApplicationDelegate is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fba8) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110c00030). One of the two will be used. Which one is undefined.
objc[45644]: Class QNSColorPanelDelegate is implemented in both /Users/abique/develop/clap-examples/builds/ninja-vcpkg/host/Debug/clap-host (0x103a6fbf8) and /Users/abique/develop/clap-examples/builds/ninja-vcpkg/plugins/Debug/clap-plugins.clap/Contents/MacOS/clap-plugins (0x110c00080). One of the two will be used. Which one is undefined.
(and so on...)
```

you should probably deactivate automatic plugin linkage for your plugin instead of adding unnecessary namespaces

**Comment by abique** [comment]
> you should probably deactivate automatic plugin linkage for your plugin instead of adding unnecessary namespaces

The plugin can't assume anything about the host. This one may very well be using Qt or not.
I don't think that I can get rid of `QNSApplication`. Can I?

I think `QT_NAMESPACE` exists specially for this type of situation: you implement a plugin statically linked to Qt and you want to avoid symbol clash.

**Comment by Neumann-A** [comment]
> you implement a plugin statically linked to Qt

You don't link plugins to plugins which in the static case really means pulling in all the symbols of one plugin into another. 

You link plugins to executable and nothing else. 

Deactivating plugins is done via setting/adding the target property `QT_NO_PLUGINS` to the target which should not link plugins.

**Comment by abique** [comment]
> You don't link plugins to plugins which in the static case really means pulling in all the symbols of one plugin into another.
> 
> You link plugins to executable and nothing else.

If you want to produce a self-contained DSO and Qt is built statically, then yes you link to the plugins as well.
What you missed I think is that the symbol clash only appears with objective-c. I suppose that objc has its own namespace / symbol table.

**Comment by JackBoosY** [comment]
If this namespace is not supported by the upstream, we may not accept this.

**Comment by abique** [comment]
https://wiki.qt.io/Qt_In_Namespace
@JackBoosY does it answer it?

**Comment by JackBoosY** [comment]
@Neumann-A What do you think about this?

**Comment by abique** [comment]
> I still think this should currently be simply an overlay

Do you mean that I can add a configure option to qtbase from a triplet overlay? Or would it imply to maintain a full copy of the port just for this namespace thing?

> speaks about different versions of qt and dlopen() (not the same version and static libs)

You missed the point. The plugin interface is host and plugin vendor neutral. Anyone can implement an host or a plugin. No assumptions can be made about what libs are being used by the host. And the host is actually using `dlopen()`, and can load multiple plugins in the same process. Also `Qt 6.2.0` and `Qt 6.2.1` are different versions, even `Qt 6.2.0` and `Qt 6.2.0` may be incompatible if compiled with different compilers, options, ...

> "self-contained" would require PRIVATE linkage against Qt not PUBLIC linkage.

I think we have a different understanding of what `PRIVATE` and `PUBLIC` linkage does.

As far as I understand it, this does not affect the exported symbols. It affects which libraries are being pulled during the link phase of another target depending on the previous one.

On linux, I will make every symbols private except my plugin entry by using: `target_link_libraries(clap-plugins PRIVATE -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/linux-clap-plugins.version)`.
On macos, I'll use: `target_link_options(clap-plugins PRIVATE -exported_symbols_list ${CMAKE_CURRENT_SOURCE_DIR}/macos-symbols.txt)`.

I'll explain again the problem with objective-c. Whatever symbol visibility you've set, the objc's runtime will register the objc's classes into a **flat namespace**. Which means that within a single objc program, you can't have two objc classes with the same name. The only solution is to use prefix/namespace to avoid name clash.

If you look at https://github.com/qt/qtbase/blob/dev/config_help.txt#L93 you'll see that `-qtnamespace` is actually a first class citizen of Qt's configure script, and they went through the pain of maintaining it since Qt4. Which shows how important that feature is.

We can continue to argue if what I'm doing is correct or not, and honestly if you can provide a better solution I'd be happy to try it. But here are some facts:
- I've built the package with a namespace and it solved my issue
- If Qt has this option and maintains it for years, then it **implies** that it is **useful**. They would not be going through this pain otherwise.

**Comment by Neumann-A** [comment]
>  Also Qt 6.2.0 and Qt 6.2.1 are different versions, even Qt 6.2.0 and Qt 6.2.0 may be incompatible if compiled with different compilers, 

unlikely. PIMPL and bin compat is the strong suite of Qt. 

> I think we have a different understanding of what PRIVATE and PUBLIC linkage does.
As far as I understand it, this does not affect the exported symbols. It affects which libraries are being pulled during the link phase of another target depending on the previous one.

Your understanding is incomplete. PRIVATE: build requirement; INTERFACE: usage requirement; PUBLIC: both.
How can something be a build requirement only? 
Simple, only use the provided stuff in your private headers and sources. This effectively hides away your usage of whatever you link and keeps it out of the link line of consumers, if you are building shared libs (If you build static libs you will still transitively require the linked libs on the link line and cmake will do that for you). As soon as you use something in a public header it is going to be PUBLIC. 


> I'll explain again the problem with objective-c. Whatever symbol visibility you've set, the objc's runtime will register the objc's classes into a flat namespace. Which means that within a single objc program, you can't have two objc classes with the same name. The only solution is to use prefix/namespace to avoid name clash.

So you are building Qt6 two times? One time with namespaces and one time without? vcpkg does not have an inconsistent view on stuff so that is really not needed here. 

 > Which shows how important that feature is.

The given reason is invalid for that conclusion.

> If Qt has this option and maintains it for years, then it implies that it is useful. They would not be going through this pain otherwise.

Yeah it is useful for vendoring Qt. E.g. If you have a superbuild of paraview and your internal Qt version is different from most systems. Then you can simply vendor Qt yourself add a pv libinfix and do the namespace stuff to get different mangled names. This make sure that your vendored version of Qt gets pulled in and avoids problems with different Qt versions. But that is not the point here.

**Comment by abique** [comment]
> > Also Qt 6.2.0 and Qt 6.2.1 are different versions, even Qt 6.2.0 and Qt 6.2.0 may be incompatible if compiled with different compilers,
> 
> unlikely. PIMPL and bin compat is the strong suite of Qt.

I'll give you one scenario:
- host happen to provide symbols for QtCore only
- plugin links with both QtCore and QtWidgets, but happens to use QtCore from the host (thanks to the dynamic linker) and QtWidgets from the plugin

If the host uses Qt 6.2 and the plugin Qt 6.3, I would not be so sure that everything is going to be fine.
Also if it happens that some classes in QtWidgets uses the internal parts of QtCore (they call it [d-pointer](https://wiki.qt.io/D-Pointer)), this one maybe has a different layout depending on the version or the compiler options. I haven't read all the code so I may be wrong here but for sure I would not bet that the compatibility is guaranteed.

> > I'll explain again the problem with objective-c. Whatever symbol visibility you've set, the objc's runtime will register the objc's classes into a flat namespace. Which means that within a single objc program, you can't have two objc classes with the same name. The only solution is to use prefix/namespace to avoid name clash.
> 
> So you are building Qt6 two times? One time with namespaces and one time without? vcpkg does not have an inconsistent view on stuff so that is really not needed here.

I'd make one triplet for the host and one for the plugins so they can work together without clashing. Consider that the host and plugins are made by different vendors, and when you're going to sell the plugin to a customer, you want to be 100% sure that it won't clash with the host or another plugin, and you want your plugin to be a single file that contains all its dependencies.

**Comment by dg0yt** [comment]
> I haven't read all the code so I may be wrong here but for sure I would not bet that the compatibility is guaranteed.

Qt is one of the most strict projects with regard to ABI compatibility. However, there are limitations.

I intentionally kept my points short. Offering namespaced Qt via triplet variables is reasonable. However, given that it won't be tested in vcpkg CI, there might be good reasons to be careful with adding this behaviour to the port.

**Comment by abique** [comment]
> I intentionally kept my points short. Offering namespaced Qt via triplet variables is reasonable. However, given that it won't be tested in vcpkg CI, there might be good reasons to be careful with adding this behaviour to the port.

If that helps I can write additional documentation in the port file, and I can add some logging if the QT_NAMESPACE is being used.

**Comment by alcroito** [comment]
Qt offers a QT_NAMESPACE define because some users need it. 
I'm not sure what's the correct vcpkg mechanism to expose it, but I don't see why it shouldn't be made available in some form. 

And as for CI testing, that's no different from not testing the various combinations of features that the port is already exposing.

**Comment by JackBoosY** [comment]
cc @ras0219-msft

**Comment by abique** [comment]
It seems that the `x86_windows` failure is due to a vcpkg crash and is unrelated to this PR.

File: ports/qtbase/portfile.cmake [hunk 1]
```diff
@@ -79,6 +79,10 @@ INVERTED_FEATURES
 list(APPEND FEATURE_OPTIONS -DCMAKE_DISABLE_FIND_PACKAGE_Libudev:BOOL=ON)
 list(APPEND FEATURE_OPTIONS -DFEATURE_xml:BOOL=ON)
 
+if(NOT QT_NAMESPACE STREQUAL "")
```
Comment (ras0219-msft) [inline ports/qtbase/portfile.cmake:82]: This does not correctly handle `QT_NAMESPACE` being undefined. For example,
```
$ cat cmaketest.cmake
if(NOT ABC STREQUAL "")
    message("bang")
endif()
$ cmake -P cmaketest.cmake
bang
```
This needs to be

```suggestion
if(QT_NAMESPACE)
```

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Other than the suggested change, I'm happy with this implementation.

LGTM, thanks for the PR!

**Comment by abique** [comment]
> Other than the suggested change, I'm happy with this implementation.
> 
> LGTM, thanks for the PR!

Thank you very much @ras0219-msft
I've applied your requested change.

**Comment by JackBoosY** [comment]
Note that qt should behave the same with and without this modification, so this changes will not break the user's expected.

**Comment by abique** [comment]
@JackBoosY the issue is flagged has require author's response but I think provided all responses already. I've also applied the suggested change.

Should we move forward?

**Comment by ras0219-msft** [comment]
Thanks!

>Note that qt should behave the same with and without this modification, so this changes should not break the user's expected.

Under most circumstances, this is exactly the consideration we should have. However, there is always the need for "if the advanced user knows what they are doing, they can tickle XYZ and they're on their own". Options that must be manually set in a custom triplet are sufficiently obscure that they may violate this (and a user that uses it is largely on their own).

**Comment by Neumann-A** [comment]
Still this is a global variable and thus needs to be prefixed with `VCPKG_` according to the vcpkg cmake guidelines. As such I have renamed it in the Qt 6.2.3 PR.

**Comment by abique** [comment]
Thank you very much everyone.

**Pull request merged**